crates/xtask/src/changelog.rs

```diff
@@ -0,0 +1,377 @@
+    ensure!(
+        std::fs::exists(&crate_path).expect("Cannot open path right now."),
+        "Crate does not exist: {}",
+        crate_path
+    );
+
+    let _changelog = read_changelog_file(crate_path)?;
+    // TODO: act upon changelog...
+
+    Ok(())
+}
+
+fn read_changelog_file(crate_path: String) -> Result<Changelog> {
+    let changelog_path = format!("{crate_path}/CHANGELOG.md");
+
+    let changelog_contents = std::fs::read_to_string(&changelog_path)
+        .with_context(|| format!("reading {changelog_path}"))?;
+
+    parse_changelog(changelog_contents)
+}
+
+/// Converts raw file contents into a Changelog data structure.
+fn parse_changelog(contents: String) -> Result<Changelog> {
+    let mut changelog = Changelog::default();
+
+    // Trim off the trailing comment: <!-- Increment to skip ... -->
+    let contents: Vec<&str> =
+        contents.split("\n\n<!-- Increment to skip CHANGELOG.md test:").collect();
+
+    // Populate increment_count if we have one.
+    if contents.len() > 1 {
+        let ending = contents[1]
+            .trim()
+            .strip_suffix("-->")
+            .expect("Malformed 'Increment to skip CHANGELOG.md test'")
+            .trim();
+
+        changelog.set_increment_count(ending.parse::<u32>().unwrap_or(0))
+    }
+
+    let contents = contents[0];
+
+    // Split by release.
+    // Skip(1) to skip the prefix.
+    for each_release in contents.split("\n## ").skip(1) {
+        let mut version_iter = each_release.split("\n### ").into_iter();
+        let release_string = version_iter.next().expect("No release version detected.").trim();
+        let mut release = ChangelogRelease::from(release_string.to_string());
+
+        // Each 'Major', 'Minor', 'Patch' section within this release.
+        for each_version in version_iter {
+            let version_parts: Vec<_> = each_version.split_terminator("\n\n").collect();
+
+            ensure!(version_parts.len() == 2, "Failed to parse release: {}", each_version);
+
+            let version_type = ChangelogReleaseType::from_str(version_parts[0], true).expect(
+                format!(
+                    "Failed to parse version_type: {}. Must be 'Major', 'Minor', or 'Patch'.",
+                    version_parts[0]
+                )
+                .as_str(),
+            );
+            let version_contents = version_parts[1]
+                .split("- ")
+                .map(|str| str.trim().to_string())
+                .filter(|str| !str.is_empty())
+                .collect();
+
+            match version_type {
+                ChangelogReleaseType::Major => release.set_major(version_contents),
+                ChangelogReleaseType::Minor => release.set_minor(version_contents),
+                ChangelogReleaseType::Patch => release.set_patch(version_contents),
+            }
+        }
+
+        changelog.push_release(release);
+    }
+
+    Ok(changelog)
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Ideally, we should use fuzzing to make sure parsing and rendering round-trips. Positive tests are not really needed because the repository is already full of changelog files, so CI will always execute parsing on those files. And negative tests are hard to write, which is why fuzzing would be simpler.

Reply:
Are you suggesting to test at a higher level? i.e. assert Changelog::parse({input}).render() == {input}

And use the existing CHANGELOG.md files?

Reply:
Yes, except that parsing may fail, which counts as a successful test. For the existing changelog, no need to do anything special, except possibly use them to bootstrap a corpus.

Reply:
Gotcha, if you are okay with it, I'll keep some small tests around. I find them helpful when iterating on logic :)